irqbalance.c

```diff
@@ -308,13 +308,20 @@ gboolean scan(gpointer data __attribute__((unused)))
 void get_irq_data(struct irq_info *irq, void *data)
 {
 	char **irqdata = (char **)data;
-	if (!*irqdata)
+	char *old = NULL;
+
+	if (!*irqdata) {
 		*irqdata = calloc(24 + 1 + 11 + 20 + 20 + 11, 1);
-	else
-		*irqdata = realloc(*irqdata, strlen(*irqdata) + 24 + 1 + 11 + 20 + 20 + 11);
+	} else {
+		old = *irqdata;
+		*irqdata = realloc(old, strlen(old) + 24 + 1 + 11 + 20 + 20 + 11);
+	}
 
-	if (!*irqdata)
+	if (!*irqdata) {
+		/* if realloc fail, not free the original memory */
+		*irqdata = old;
 		return;
+	}
```

Comment:
You can make this far simpler. Just declare a new void pointer and make that store the return value of calloc/realloc, then check that, return if NULL, and in the fall through case, assign the result to *irqdata

 void get_irq_data(struct irq_info *irq, void *data)
 {
        char **irqdata = (char **)data;
+       void *newptr;
        if (!*irqdata)
-               *irqdata = calloc(24 + 1 + 11 + 20 + 20 + 11, 1);
+               newptr = calloc(24 + 1 + 11 + 20 + 20 + 11, 1);
        else
-               *irqdata = realloc(*irqdata, strlen(*irqdata) + 24 + 1 + 11 + 20 + 20 + 11);
+               newptr = realloc(*irqdata, strlen(*irqdata) + 24 + 1 + 11 + 20 + 20 + 11);
 
+       if (!newptr)
+               return;
+       *irqdata = newptr;
+

Reply:
ok, I have modify the patch. thanks.